packages/core/contracts/financial-templates/funding-rate-store/implementation/FundingRateStore.sol

```diff
@@ -49,21 +52,53 @@ contract FundingRateStore is FundingRateStoreInterface, Testable, Lockable {
     mapping(address => mapping(uint256 => FundingRateRecord)) public fundingRateDisputes;
 
     uint256 public proposalLiveness;
+
     /****************************************
      *                EVENTS                *
      ****************************************/
 
-    event ProposedRate(address indexed perpetual, int256 rate, uint256 proposalTime, address indexed proposer);
+    event ChangedRewardRate(address indexed perpetual, uint256 rewardRate);
+    event ProposedRate(
+        address indexed perpetual,
+        int256 rate,
+        uint256 indexed proposalTime,
+        address indexed proposer,
+        uint256 rewardPct
+    );
     event DisputedRate(
         address indexed perpetual,
         int256 rate,
-        uint256 proposalTime,
+        uint256 indexed proposalTime,
         address indexed proposer,
-        address indexed disputer
+        address disputer
+    );
+    event PublishedRate(
+        address indexed perpetual,
+        int256 rate,
+        uint256 indexed proposalTime,
+        address indexed proposer,
+        uint256 rewardPct,
+        uint256 reward
+    );
+    event DisputedRateSettled(
+        address indexed perpetual,
+        uint256 indexed proposalTime,
+        address proposer,
+        address indexed disputer,
+        bool disputeSucceeded
     );
-    event PublishedRate(address indexed perpetual, int256 rate, uint256 proposalTime, address indexed proposer);
     event FinalFeesPaid(address indexed collateralCurrency, uint256 indexed amount);
 
+    /****************************************
+     *                MODIFIERS             *
+     ****************************************/
+
+    // Pubishes any pending proposals whose liveness has passed, pays out rewards for such proposals
+    modifier publishAndWithdrawProposal(address perpetual) {
```

Comment:
This internal method, and modifier added to every external contract call, is the bulk of this contract's changes. After discussions with @mrice32, we decided that there should be an external method callable by anyone to withdraw proposal rewards.

Additionally, every method ultimately was checking whether a pending proposal should be published, so I decided to move this into an internal method and have every method call it before running their core logic.